[InstCombine] Match intrinsic recurrences when known to be hoisted

antoniofrighetto opened the pull request:
For value-accumulating recurrences of kind:
```
  %umax.acc = phi i8 [ %umax, %backedge ], [ %a, %entry ]
  %umax = call i8 @llvm.umax.i8(i8 %umax.acc, i8 %b)
```
The binary intrinsic may be simplified into an intrinsic with init
value and the other operand, if the latter is loop-invariant:
```
  %umax = call i8 @llvm.umax.i8(i8 %a, i8 %b)
```

Proofs: https://alive2.llvm.org/ce/z/ea2cVC.

Fixes: https://github.com/llvm/llvm-project/issues/145875.

antoniofrighetto commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1532,6 +1532,46 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
+  // A binary intrinsic recurrence with loop-invariant operands is equivalent to
+  // `call @llvm.binary.intrinsic(Init, OtherOp)`.
+  if (!matchSimpleBinaryIntrinsicRecurrence(II, PN, Init, OtherOp) ||
+      isa<Constant>(OtherOp) || !PN->hasOneUse() ||
```
Looking at the test in known-phi-recurse.ll, I thought we didn't want to do this here in InstCombine. Looking better, I think it's a groundless concern. Likewise, I thought the multi-use phi may have not been profitable as could have lead to have multiple intrinsic operations around (one hoisted, one in the loop?); but I don't think it's the case, and indeed we should make the intrinsic loop-invariant. Changed, thanks.

antoniofrighetto commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1532,6 +1532,46 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
+  // `call @llvm.binary.intrinsic(Init, OtherOp)`.
+  if (!matchSimpleBinaryIntrinsicRecurrence(II, PN, Init, OtherOp) ||
+      isa<Constant>(OtherOp) || !PN->hasOneUse() ||
+      !IC.getDominatorTree().dominates(OtherOp, PN))
```
Think there doesn't exist an overload where the definition is a value and the use is a BB?

nikic replied:
Ah, right, I forgot we don't have that overload due to Value/BasicBlock ambiguity.

antoniofrighetto commented on llvm/test/Transforms/InstCombine/recurrence-binary-intrinsic.ll:
Added a test w/ copysign, thanks (think we could add that too, but I wouldn't expect to see a copysign in a recurrence?).

nikic replied:
I'd suggest using something like uadd.sat or usub.sat, where the transform would be invalid because the intrinsic is not idempotent (in the sense that `f(f(x, y), y) != f(x, y)`).

antoniofrighetto replied:
That was indeed a poor negative test, updated, thanks.

nikic reviewed (approved):
LGTM

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1532,6 +1532,41 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
+    return nullptr;
+  }
+
+  auto *InvariantBinaryInst =
+      IC.Builder.CreateBinaryIntrinsic(IID, Init, OtherOp);
+  if (isa<FPMathOperator>(InvariantBinaryInst))
+    cast<Instruction>(InvariantBinaryInst)->copyFastMathFlags(II);
```
```suggestion
  auto *InvariantBinaryInst =
      IC.Builder.CreateBinaryIntrinsic(IID, Init, OtherOp, /*FMFSource=*/II);
```

antoniofrighetto replied:
That's unfortunately not possible per assertion `isa<FPMathOperator>(this) && "getting fast-math flag on invalid op"'`  on non-fast-math binary intrinsics. Perhaps we may want to change this in IRBuilder intrinsic creation in the future.

dtcxzyw replied:
Use `dyn_cast<FPMathOperator>(II)`?

antoniofrighetto replied:
In this case, I think you'd need another `dyn_cast<Instruction>(InvariantBinaryInst)` to check that `CreateBinaryIntrinsic` didn't constant-fold, as copyFastMathFlags is private to FPMathOperator. Think the current one may be cleaner.

dtcxzyw reviewed (approved):
LG

antoniofrighetto commented:
> Hm, I'm curious whether we would get this optimization "for free" if we called foldOpIntoPhi for intrinsics.

Tried by pivoting the fold in the intrinsics visitor as follows:
```cpp
PHINode *PN;
Value *Init, *OtherOp;
if (matchSimpleBinaryIntrinsicRecurrence(II, PN, Init, OtherOp) &&
    DT.dominates(OtherOp, PN))
  if (auto *Res = foldOpIntoPhi(*II, PN, true))
    return Res;
```
We fold correctly the one-use phi case, however for the multiuse we bail out in: https://github.com/llvm/llvm-project/blob/c9ceb9b75fd547c7d2e79837075370f4c8db8faa/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L1947-L1948

May as well take a deeper look as a follow-up patch, if you say.

antoniofrighetto commented:
> The question then would be whether handling multi-use is worth a dedicated fold. (Though possibly we could extend foldOpIntoPhi to handle this specific multi-use case, not sure...)

Comparing the results of the first run https://github.com/dtcxzyw/llvm-opt-benchmark/pull/2596 (w/o multi-use phi) and the one in https://github.com/llvm/llvm-project/pull/151115, and it seems like there are four test cases that were not folded via foldOpIntoPhi. Wouldn't this imply that #151115 does not entirely subsume this fold?

nikic commented:
Can you please rebase over https://github.com/llvm/llvm-project/pull/151115? Then we can get a clean diff from llvm-opt-benchmark.

antoniofrighetto commented:
From the diff, it looks like the fold got triggered only with multi-use phi (though this folds fmaximumnum/fminimumnum too). Could make sense rooting this after foldOpIntoPhi with !PN->hasOneUse()?

I took a look at extending foldOpIntoPhi; though, if I didn't miss anything, it didn't look that promising unless passing an extra, say, IdempotentBinaryIntrinsic boolean or including the fold inside `if (!OneUse && !IdenticalUsers)` (which none of the two look that elegant to me).

antoniofrighetto commented:
Am assuming this patch is still fine to merge as it is.

antoniofrighetto committed:
[InstCombine] Match intrinsic recurrences when known to be hoisted

For value-accumulating recurrences of kind:
```
  %umax.acc = phi i8 [ %umax, %backedge ], [ %a, %entry ]
  %umax = call i8 @llvm.umax.i8(i8 %umax.acc, i8 %b)
```
The binary intrinsic may be simplified into an intrinsic with init
value and the other operand, if the latter is loop-invariant:
```
  %umax = call i8 @llvm.umax.i8(i8 %a, i8 %b)
```

Proofs: https://alive2.llvm.org/ce/z/ea2cVC.

Fixes: https://github.com/llvm/llvm-project/issues/145875.
